### ISSUE TITLE
Eventgen flooding me.

### ISSUE THREAD

**Issue by memarshall63** [splunk/eventgen#89]
I've tried for a couple days to figure out what's gone wrong with my setup, and just need some expert help.  My issue is on both Windows and a MAC install.  I just can get Eventgen to slow down the messages it sends. 

My Eventgen.conf looks like this:

> [sample.tutorial1]
> mode = replay
> sampletype = csv
> timeMultiple = 2
> interval=10
> #backfill = -15m
> #backfillSearch = index=main sourcetype=splunkd
> 
> outputMode = splunkstream
> splunkHost = localhost
> splunkUser = admin
> splunkPass = <password>
> 
> { token lines omitted}

I must be missing a setup or something --  it just kicks out about 200 messages in the same second and floods me.  I looked at enough of the python to know that it's in the Timer routine.  I just can't get it to sleep between messages.   Interval?, timeMultiple? Replay mode?    I even tried 'sample' mode, and that didn't seem to change anything either.  I have about 8 pages of this message all in the same second.   

> 2018-10-14 12:27:50 | Mikes-MacBook-Air.local | _internal | 2018-10-14 12:27:50 eventgen        INFO     MainProcess Self.sample.interval = '0'
> -- | -- | -- | --
> 2018-10-14 12:27:50 | Mikes-MacBook-Air.local | _internal | 2018-10-14 12:27:50 eventgen        INFO     MainProcess Worker# 0: Put 1 events in queue for sample 'sample.tutorial1' with et '2018-10-14 12:27:50.843158' and lt '2018-10-14 12:27:50.843173'
> 2018-10-14 12:27:50 | Mikes-MacBook-Air.local | _internal | 2018-10-14 12:27:50 eventgen        INFO     MainProcess Start '1' generatorWorkers for sample 'sample.tutorial1'

My sample file has 4 messages in it.  (The first 4 from the 'sample.tutorial1').

The only mod (besides .conf files and sample files) I've made is I did have to comment the SIGPIPE signal for the Windows install.

What goofball mistake have I made?  Any help would be appreciated.
-Mike.

**Comment by lephino** [comment]
shut off debug in the modinput

**Comment by arctan5x** [comment]
Interval is used to give pause between events. Would you post your eventgen.conf and sample files so we can investigate it? Also, specifying end=<int> will give a stop to your data generation after <int> iterations.

**Comment by memarshall63** [comment]
Thanks.   I've re-loaded the latest SA-Eventgen, and reset my eventgen.conf and sample files.  I enabled the modinput, and....     I don't get any events generated, but I get a repeating log message.  Attached are my eventgen.conf, samples file, and a csv of the output of [[ index=_internal sourcetype="eventgen_main" OR sourcetype="eventgen_modinput" | table _time, host, index, _raw ]]

Appreciate any help you can give.  

[eventgen.conf.txt](https://github.com/splunk/eventgen/files/2564515/eventgen.conf.txt)
[sample.mobilemusic_mm.csv.txt](https://github.com/splunk/eventgen/files/2564516/sample.mobilemusic_mm.csv.txt)
[log1541734755_211.csv.txt](https://github.com/splunk/eventgen/files/2564537/log1541734755_211.csv.txt)

**Comment by jmeixensperger** [comment]
Where is your eventgen.conf and sample located in the filesystem?

**Comment by memarshall63** [comment]
eventgen.conf is in "local" beneath my eventgentest application.  sample is in "samples" at the same level as "local".   I did have the eventgentest index configured.  I cleaned up the eventgen.conf (there was a bad cr/lf in it..) and tried again.  Pretty much same results.  However, I left the thing run for a couple hours and watch the index.   And.. eventually, it did put messages into the index.  It put the same message in the index 120 times in 1 second... waited about 5 mins.. did it again.  Did it 6 times... then waited for another hour... and started the same process over.    I've attached the exported _internal and eventgentest indexes from the timeframes.    I'm just not sure what's wrong.  Possibly, if you could point me to a working sample, I could just start over and get something working and go from there?   Thanks for the assistance.   I do appreciate it.


![screen shot 2018-11-24 at 6 28 18 pm](https://user-images.githubusercontent.com/33108108/48973855-583b5e80-f017-11e8-900d-c433527992af.png)
![screen shot 2018-11-24 at 6 28 33 pm](https://user-images.githubusercontent.com/33108108/48973858-5ffb0300-f017-11e8-8200-ec34f37bbc6c.png)

[eventgen_index.csv.txt](https://github.com/splunk/eventgen/files/2612675/eventgen_index.csv.txt)
[_internal_eventgen.csv.txt](https://github.com/splunk/eventgen/files/2612676/_internal_eventgen.csv.txt)

**Comment by jmeixensperger** [comment]
I tried to reproduce with your conf and sample files, but I am still getting a steady stream of events in the "eventgentest" index. My only change was to splunkPassword to make sure authentication worked. I would suggest retrying the install/generation on a fresh splunk instance. You can also try to use the same testing application that I did: "Cisco Networks Add-on for Splunk Enterprise" (careful--there are 2 identically-named apps, choose the second one). If you are still having trouble, maybe we can set up a one-on-one to troubleshoot.

**Comment by memarshall63** [comment]
Hi, thanks for the help!   Could you expand a little on "a steady stream of events"?   I'm just wondering what this should look like "normally".   This weekend, I'll tear things down and try to build it back up again.

**Comment by jmeixensperger** [comment]
Sure, I let it generate for around 15 minutes and here's what my search results look like:

<img width="1698" alt="screen shot 2018-11-29 at 5 40 07 pm" src="https://user-images.githubusercontent.com/16170719/49263094-f6576c00-f3fd-11e8-8914-a28b4ddb9e07.png">


Here's the eventgen.conf I used too if you want to double check yours against it (I've removed my password):
[eventgen.conf.txt](https://github.com/splunk/eventgen/files/2631621/eventgen.conf.txt)

**Comment by memarshall63** [comment]
Working on this attempt..  I put my eventgen.conf and my samples into a freshly downloaded copy of Cisco networks add-in..  I put my sample in the samples directory and my .conf file in "local".   There's the shipped eventgen.conf in default -- but I left that alone. Go to my modinput and enable it... and I see it start in the internal logs.. it puts out the cisco sample messages... I restarted Splunk... and it picked up my messages, but still exhibits the same 120 messages all at once.    The cisco sample seems "normal" though.. There are 77 messages in the sample.. and it nicely spits 77 messages in over  about 10 minutes.  Cisco using sample mode.  I'm using Replay mode.   

Let me twist on it awhile.. I have one working and one not.. so I can start to tear them down bit by bit.
Thanks for the assist.
![screen shot 2018-12-04 at 10 51 03 pm](https://user-images.githubusercontent.com/33108108/49489016-34e29180-f817-11e8-9f47-befa1e419002.png)

**Comment by jmeixensperger** [comment]
I'm glad to hear that you have a sample working. As far as the 120 messages in a second, what version of eventgen are you using? Are they the same log messages that you provided in your original message?

**Comment by mossytrees** [comment]
@memarshall63 I just ran into a similar problem.  I was using `mode = replay` with a sample file with 36 entries.  I had `interval` set to 300 and was backfilling for 24 hours, so I was expecting 10,368 events to be generated.  (`24*60/5*36 = 10,368`).  Instead, I got hundreds of thousands of events being generated that spanned less than an hour.  I killed splunk and went to the docs and noticed:

```
interval = <integer>
    * Only valid in mode = sample
```

Ok, so interval won't apply in replay mode... time to consult the `mode` docs to see how `replay` figures out how often to generate events:

```
mode = sample | replay
    * Default is sample, which will generate count (+/- rating) events every configured interval
    * Replay will instead read the file and leak out events, replacing timestamps
```

Ok, a little helpful... but when I found this, something clicked:

```
timeMultiple = <float>
    * Only valid in mode = replay
    * Will slow down the replay of events by <float> factor.  For example, allows a 10 minute sample
      to play out over 20 minutes with a timeMultiple of 2, or 60 minutes with a timeMultiple of 6.
      By the converse, make timeMultiple 0.5 will make the events run twice as fast.
```

Looks like it uses the sample file's rate of events (modified by `timeMultiple` if set) to determine how fast to generate events.

Back to my sample file:  all 36 events have the exact same timestamp.  I noticed you said you're using just the first 4 lines of `sample.tutorial1`, which also have all just 1 timestamp.  I can only assume that in this case eventGen behaves as if your events are continuous and generates events as fast as it can.

I cleaned my index (`splunk clean eventdata main`), and switched to `mode = sample`.  As soon as I did that and booted up Splunk again I had my 10,368 events as expected.  I see above you mentioned trying `sample` mode as well; are you certain you weren't seeing leftover events from your previous attempt that generated all the extra events??  (You can try using a "realtime" over all time search to see what events are currently being indexed, or just clean your index and start over to be sure.)

**Comment by jmeixensperger** [comment]
@mossytrees The fix for the get_time_difference error is included #121 and will be in the next release.
